util/eventbus/bus.go

```diff
@@ -173,6 +180,27 @@ func (b *Bus) shouldPublish(t reflect.Type) bool {
 	return len(b.topics[t]) > 0
 }
 
+func (b *Bus) listClients() []*Client {
+	b.clientsMu.Lock()
+	defer b.clientsMu.Unlock()
+	return b.clients.Clone().Slice()
+}
+
+func (b *Bus) snapshotPublishQueue() []PublishedEvent {
+	resp := make(chan []PublishedEvent)
+	select {
+	case b.snapshot <- resp:
+		select {
+		case snapshot := <-resp:
```

Comment:
I think as written this can maybe deadlock (or if not, then I think the done check is unnecessary): `pump` receives our channel, grabs a snapshot, and blocks waiting for this goroutine to rendezvous here. But if the context is cancelled and we see the router as done, we will never receive, locking the pump goroutine in a branch of its select that can never terminate.

On the other hand, if the router ended earlier, our send could never have succeeded in the first place: Once pump has our channel it won't look again at its stop signal till it's delivered us a value.

So I think we not only don't _need_ to check for done-ness here, we maybe _shouldn't_.

Reply:
Well spotted! As you say, once the snapshot request is accepted, the response will come unconditionally and we must not block it. Fixed, here and in the subscribe path.